### ISSUE TITLE
extension.node-debug.startSession failed in some environments

### ISSUE THREAD

**Issue by dmblack** [Microsoft/vscode#23965]
**Root Cause:**
nvm (https://github.com/creationix/nvm) provides node via .bashrc, which these launchers to not invoke.

**Resolution / Workaround:**
Launch code with exposed path for node.

In my case; via gnome-terminal (Which loads .bashrc)

- VSCode Version: 
Version 1.11.0
Commit 9122ec4b1068bd391e41974f46a4b54c1b9c90f9
Date 2017-04-05T21:13:24.700Z
Shell 1.4.6
Renderer 53.0.2785.143
Node 6.5.0

- OS Version:
debian_version
8.7

Steps to Reproduce:
1. Launch `code` via Run Program / Quick Launch (ALT + F2)
2. Attempt to use any form of Debugger

Error Message: 
Running the contributed command:'extension.node-debug.startSession' failed.

(edited)
Ensuring node is available in your ENV, and launching from terminal is a valid workaround.
`code`

Feel free to close this issue if it's expected behavior. This issue was logged as search results for the error message are non-existent.

**Comment by chrmarti** [comment]
When you launch through Run Program, do you open VSCode on a folder or do you just open single files?

Is there anything in the devtools' console (Help > Toggle Developer Tools)?

**Comment by dmblack** [comment]
@chrmarti 
Thanks for your response!

Via ALT + F2, (Or the 'linux quick launch') I'm only typing;
`code`

Developer tools provides these errors;

```
messageService.ts:126 Running the contributed command:'extension.node-debug.startSession' failed.e.doShow @ messageService.ts:126e.show @ messageService.ts:105(anonymous function) @ compositePart.ts:298r @ eventEmitter.ts:315e._emitToSpecificTypeListeners @ eventEmitter.ts:182e._emitEvents @ eventEmitter.ts:201e.emit @ eventEmitter.ts:216(anonymous function) @ actions.ts:232v @ winjs.base.raw.js:1209enter @ winjs.base.raw.js:901_run @ winjs.base.raw.js:1068_error @ winjs.base.raw.js:1041n @ winjs.base.raw.js:743v @ winjs.base.raw.js:1209enter @ winjs.base.raw.js:901_run @ winjs.base.raw.js:1068_error @ winjs.base.raw.js:1041e.resolveErr @ ipcRemoteCom.ts:111f @ ipcRemoteCom.ts:203_combinedTickCallback @ internal/process/next_tick.js:67_tickCallback @ internal/process/next_tick.js:98
shell.ts:475 Running the contributed command:'extension.node-debug.startSession' failed.e.onUnexpectedError @ shell.ts:475(anonymous function) @ shell.ts:435e.onUnexpectedError @ errors.ts:68o @ errors.ts:88(anonymous function) @ winjs.base.js:43(anonymous function) @ winjs.base.js:38
winjs.base.js:45 WARNING: Promise with no error callback:82
winjs.base.js:46 Object {exception: null, error: "Running the contributed command:'extension.node-debug.startSession' failed.", promise: n.C…s.d…e._oncancel, handler: undefined, id: 82…}
```

Note:
With additional configuration property;
`"protocol": "inspector",`
The error changes to;
> Cannot find runtime 'node' on PATH.

**Comment by roblourens** [comment]
Do you have node installed? You have to have node on your path to use the node debugger.

**Comment by dmblack** [comment]
@roblourens 
Thanks for your response.

I do have node installed, via nvm (https://github.com/creationix/nvm). Hence, launching via `code .` is not a problem.

**Comment by dmblack** [comment]
@roblourens @chrmarti 
Note: I've updated the initial report.

You can avoid this issue via ALT + F2 if you provide `code .`.

**Comment by roblourens** [comment]
And if you type `node` in the vscode integrated terminal, does it run node successfully?

**Comment by roblourens** [comment]
What does alt+f2 do? I don't see a default keybinding for that.

**Comment by dmblack** [comment]
@roblourens 
Thanks again for your response!

Yes. But I think the integrated terminal is literally spawning a new 'session'. ENV variables are available.

**Comment by roblourens** [comment]
Can you try launching with `"protocol": "inspector"`?

**Comment by dmblack** [comment]
@roblourens 
ALT + F2 is a bit like Start -> Run (Win + R) in Windows. It's just a launcher.

When invoked via fluxbox, it starts 'fbrun'. But this issue happens with other desktop environments also (xfce, lxde, cinnamon).

**Comment by dmblack** [comment]
@roblourens 
I attempted this in an earlier note, targeted at chrmarti;

 With additional configuration property;
"protocol": "inspector",
The error changes to;
> Cannot find runtime 'node' on PATH.

**Comment by roblourens** [comment]
Ok, so if you launch code with `code .`, then debugging works, but if you launch `code` then open a folder, you see "Cannot find runtime 'node' on PATH".

**Comment by dmblack** [comment]
@roblourens 
Correct.

If I do not have the protocol property provided at all, error message is again different.

**Comment by dmblack** [comment]
Further Information

**Process without path:**
`/usr/share/code/code`

`/usr/share/code/code --type=zygote --no-sandbox`

`/usr/share/code/code --type=renderer --no-sandbox --primordial-pipe-token=E20147C04E103196FD32D6A556E22820 --lang=en-GB --node-integration=true --background-color=#1E1E1E --enable-pinch --num-raster-threads=2 --content-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --video-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --disable-accelerated-video-decode --disable-webrtc-hw-encoding --disable-gpu-compositing --mojo-channel-token=DD799CC36AFCDCA401CE988E6449B9A4 --mojo-application-channel-token=E20147C04E103196FD32D6A556E22820 --channel=20568.0.680731602 --v8-natives-passed-by-fd --v8-snapshot-passed-by-fd`

`/usr/share/code/code --type=renderer --no-sandbox --primordial-pipe-token=FFF0536404471ABFF73F82C9525810C3 --lang=en-GB --node-integration=true --hidden-page --enable-pinch --num-raster-threads=2 --content-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --video-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --disable-accelerated-video-decode --disable-webrtc-hw-encoding --disable-gpu-compositing --mojo-channel-token=A837918889E2BDF363C3B11F50D68012 --mojo-application-channel-token=FFF0536404471ABFF73F82C9525810C3 --channel=20568.1.624949248 --v8-natives-passed-by-fd --v8-snapshot-passed-by-fd`

`/usr/share/code/code ./terminalProcess`

`/usr/share/code/code /usr/share/code/resources/app/out/bootstrap --type=extensionHost`

`/usr/share/code/code /usr/share/code/resources/app/out/bootstrap --type=watcherService`

`/usr/share/code/code /usr/share/code/resources/app/extensions/json/node_modules/vscode-languageclient/lib/utils/electronForkStart /usr/share/code/resources/app/extensions/json/server/out/jsonServerMain.js --node-ipc`


**Process with path:**
`/usr/share/code/code .`

`/usr/share/code/code --type=zygote --no-sandbox`

`/usr/share/code/code --type=renderer --no-sandbox --primordial-pipe-token=3309E3AD2093434EF4E48DD628450229 --lang=en-GB --node-integration=true --background-color=#1E1E1E --enable-pinch --num-raster-threads=2 --content-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --video-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --disable-accelerated-video-decode --disable-webrtc-hw-encoding --disable-gpu-compositing --mojo-channel-token=D764C2896156352D229FC18D05218C40 --mojo-application-channel-token=3309E3AD2093434EF4E48DD628450229 --channel=19991.2.1478782836 --v8-natives-passed-by-fd --v8-snapshot-passed-by-fd`

`/usr/share/code/code --type=renderer --no-sandbox --primordial-pipe-token=5021E94BE77F620B1B0D32792FCAA3F8 --lang=en-GB --node-integration=true --hidden-page --enable-pinch --num-raster-threads=2 --content-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --video-image-texture-target=3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553,3553 --disable-accelerated-video-decode --disable-webrtc-hw-encoding --disable-gpu-compositing --mojo-channel-token=9730E6429B7BEFB59A4281F121023629 --mojo-application-channel-token=5021E94BE77F620B1B0D32792FCAA3F8 --channel=19991.1.1932691197 --v8-natives-passed-by-fd --v8-snapshot-passed-by-fd`

`/usr/share/code/code ./terminalProcess`

`/usr/share/code/code /usr/share/code/resources/app/out/bootstrap --type=extensionHost`

`/usr/share/code/code /usr/share/code/resources/app/out/bootstrap --type=watcherService`

`/usr/share/code/code /usr/share/code/resources/app/extensions/json/node_modules/vscode-languageclient/lib/utils/electronForkStart /usr/share/code/resources/app/extensions/json/server/out/jsonServerMain.js --node-ipc`

**Comment by dmblack** [comment]
@roblourens @chrmarti 

Further investigations via /proc/<pid>/environ, on both types of launch, indicates that this is definitely related to node's path availability via ALT + F2.

Though fbrun is the fluxbox implementation, lxde, xfce, cinnamon, and mate all show the same issue with their respective launchers. However; the root cause is the way nvm provides node.

By design; nvm provides my node path via .bashrc. These launchers must not invoke .bashrc.

Long story short;
Combination of quick launch, and nvm (https://github.com/creationix/nvm); results in this situation.

I've updated the initial report to make this clear, and provide a result for anyone else with this issue.

**Comment by xsstomy** [comment]
same ， @dmblack   do you know how to solve it ?

**Comment by dmblack** [comment]
@xsstomy 
I've updated the initial issue report to include the steps I needed to solve it.

Briefly; my node path was not available to VSCode, due to the way NVM loads via bashrc.

**Comment by xsstomy** [comment]
@dmblack 
Thanks  for your response!

**Comment by kennasoft** [comment]
Hi @dmblack, I only started experiencing this since I updated to 1.12.1. I had been able to debug even with nvm before now. This is my first time of encountering this error.

I'm not sure I understood your workaround. I'm running vscode on MacOS Sierra (10.12.4)

**Comment by superbland** [comment]
@kennasoft I'm also experiencing this as of this morning after VSC seems to have updated. For me adding `"protocol": "lagacy"` to the launch.json as suggested by @roblourens appears to resolve this. However I'm not sure if this a a bug or expected behaviour.

MacOS El Capitan 10.11.16, VSC 1.12.1

**Comment by kennasoft** [comment]
@superbland that solves it for me too. Thanks;
I'd be surprised if it's expected behaviour, cos the documentation seems to recommend a move away from the legacy protocol.

**Comment by amaqkit** [comment]
I meet this before, the environment is OK, I fixed this by disable the Dash[https://marketplace.visualstudio.com/items?itemName=deerawan.vscode-dash] plugin.

**Comment by danielepolencic** [comment]
Same here on MacOSX 10.12.4 and the newly updated VSCode 1.12.1.

Adding  `"protocol": "lagacy"` solves the issue.

**Comment by pluddy** [comment]
Same problem here, 1.12.1 on macOS 10.12.4, 
"protocol": "legacy" solved it.

**Comment by donnrenk** [comment]
Same issue here after the latest VS code update; same solution ("protocol": "legacy" to the node launch configuration object).

**Comment by levicot** [comment]
Same problem here after the latest VS code update (1.12.1) on xubuntu; same solution ("protocol": "legacy" to the node launch configuration object).

**Comment by heitorlessa** [comment]
Same here: **problem solved by updating ``launch.json``**

**Comment by idchlife** [comment]
@levicot this works!

**Comment by z-vr** [comment]
all of the sudden it stopped working without legacy protocol. literally was all fine, then i restarted VS, and it broke

**Comment by idchlife** [comment]
@z-vr maybe you restarted VSCode and you were in the middle of debug process?

**Comment by alesju** [comment]
Same issue here. Had to apply legacy protocol for it to work, worked fine for me on Friday and stopped working over the weekend (when I didn't even turn on the laptop).

macOS 10.12.5

**Comment by dmblack** [comment]
Thanks to all the other users posting feedback on their fixes for MacOS.

I've updated the original report to include this fix - so people don't have to scroll through pages of updates.

**Comment by christyharagan** [comment]
Hi all,

Thought I'd chime in. I didn't found the answer of "when on Mac, use the legacy protocol" particularly useful, especially now Node v8.0.0 is out, which doesn't support legacy anymore (I think I'm right in saying that).

I believe the problem in my case - at least - was due to me previously having installed the experimental version of the inspector debugger (and/or the the chrome debugger, not exactly sure which). However, even though they have been removed, they seemed to leave artefacts that were interfering and causing this problem. 

So I decided to clean uninstall and reinstall, which fixed it. And this is the key part:

Simply removing ```/Applications/VS Code``` and ```~/.vscode``` was not enough. There is also a folder called ```/Users/<user name>/Library/Application\ Support/Code``` that needs to be removed. 

With this folder removed and a new install of VSCode, it's right back to factory settings. Everything works fine now.

Hoping this helps some future person, because

**Comment by ProjectInfinity** [comment]
Manually changed protocol to inspector and it worked for me on node 8.0.

**Comment by amigo421** [comment]
the issue is closed, but can someone explain me how to avoid the problem? 
vscode 1.12.2
os: mint 18.1
node 7.2.1

**Comment by k0d3d** [comment]
Guys, I had this problem too just after I installed the Wakatime vscode plugin. I disabled it and it started working .

**Comment by thomassuckow** [comment]
This issue also seems to crop up if you restart your mac and it automatically reopens VSCode for you. Restarting VSCode corrects the issue.

**Comment by JacobJT** [comment]
@dmblack Could you edit @thomassuckow 's response into the top post? While `"protocol": "legacy"` in the launch.json did allow it to work, Thomas' fix did without having to modify my code. I was having trouble finding relevant info on what that option does so am weary of modifying my code to include it if not necessary.

**Comment by weinand** [comment]
I will lock this issue since it is too time consuming to follow all the various discussions that were added after the issue has been closed.
Please create new issue and we will investigate.
